x-pack/plugins/enterprise_search/KEA.md

```diff
@@ -0,0 +1,178 @@
+# Kea State Management
+Enterprise Search uses [Kea.js](https://github.com/keajs/kea) to manage our React/Redux state for us. Kea state is handled in our `*Logic` files and exposes [values](https://kea.js.org/docs/guide/concepts#values) and [actions](https://kea.js.org/docs/guide/concepts#actions) for our components to get and set state with.
+
+To make sure we have a consistent way of dealing with state, we have a few guidelines we follow. Note that these are guidelines, not hard and fast rules, and you can deviate from them if it makes sense. In addition, there's a lot of legacy code in our codebase that does not adhere to these standards. New code is expected to follow these guidelines as much as possible.
+
+## Only use Kea when necessary
+
+Redux and Kea are excellent tools for state management in complex applications, to maintain consistent state across different components and decouple their functionality from APIs and complex application state considerations. However, overuse of state management tools can have the opposite effect of creating a highly-coupled application where any changes to the state logic have unintended consequences across the codebase. In addition, components that rely on Kea are harder to test than those that don't. Be mindful of where you apply these tools.
```

Comment:
Could you elaborate a little here on the passing from parent vs pulling from logic decision for a component?  I think we've favored pulling directly from the containing "view" logic more often than not.  Trying to avoid this I feel like I end up passing a lot of props down mutliple  nested components. You can see this in the `InlineEditableTable` family of components.

Reply:
I'll add a little bit. Basically in my mind:

- Passing down a (nested) object that is used at each intermediate step of the hierarchy: fine.
- Passing down a variable through many layers that you only end up using at the end of the chain: less fine, but can sometimes work. Kind of a 'feel' thing. 
- Passing events up or down through multiple levels: avoid like the plague, use Kea.